Services/Catalogue/AttributesManager.php

```diff
@@ -85,6 +93,13 @@ public function getSwatchDetails($product, $attribute)
 
             $swatchValue = $this->getSwatchValues($optionIds);
             if (count($optionIds) && $swatchValue !== null) {
+                if ($this->storeCatalogueConfig->shouldUseProductImageAsSwatch() && $swatchType === 'visual') {
+                    return [
+                        'type' => $swatchType,
+                        'value' => $this->productImageManager->getThumbnail($product, $product->getThumbnail()),
```

Comment:
If we already have `mainImage` as processed till here. We can directly use it instead of fetching it again.